tortoise/verifying_tortoise.go

```diff
@@ -0,0 +1,602 @@
+package tortoise
+
+import (
+	"errors"
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"github.com/spacemeshos/go-spacemesh/mesh"
+)
+
+// MaxExceptionList is the maximum number of exceptions we agree to accept in a block
+const MaxExceptionList = 100
+
+//We keep a table that records, for each block, its votes about every previous block
+//We keep a vector containing our vote totals (positive and negative) for every previous block
+//When a new block arrives, we look up the block it points to in our table,
+// and add the corresponding vector (multiplied by the block weight) to our own vote-totals vector.
+//We then add the vote difference vector and the explicit vote vector to our vote-totals vector.
+
+type blockDataProvider interface {
+	GetBlock(id types.BlockID) (*types.Block, error)
+	LayerBlockIds(l types.LayerID) (ids []types.BlockID, err error)
+
+	GetLayerInputVector(lyrid types.LayerID) ([]types.BlockID, error)
+	SaveLayerInputVector(lyrid types.LayerID, vector []types.BlockID) error
+
+	SaveContextualValidity(id types.BlockID, valid bool) error
+
+	Persist(key []byte, v interface{}) error
+	Retrieve(key []byte, v interface{}) (interface{}, error)
+}
+
+func blockMapToArray(m map[types.BlockID]struct{}) []types.BlockID {
+	arr := make([]types.BlockID, len(m))
+	i := 0
+	for b := range m {
+		arr[i] = b
+		i++
+	}
+	return types.SortBlockIDs(arr)
+}
+
+type turtle struct {
+	logger log.Log
+
+	bdp blockDataProvider
+
+	Last  types.LayerID
```

Comment:
should all members be exported?

Reply:
The exporting is needed because currently the recovery mechanism is to save all state to disk represented as XDR and it doesn't include un-exported fields.